src/mcts/search.cc

```diff
@@ -312,9 +312,16 @@ void Search::UpdateRemainingMoves() {
   if (limits_.time_ms >= 0) {
     auto time_since_start = GetTimeSinceStart();
     if (time_since_start > kSmartPruningToleranceMs) {
-      auto nps = (1000LL * total_playouts_ + kSmartPruningToleranceNodes) /
+      float nps = (1000.0f * total_playouts_ + kSmartPruningToleranceNodes) /
                      (time_since_start - kSmartPruningToleranceMs) +
                  1;
+      if (nps_ < 0) {
+        nps_ = nps;
+      } else {
+        // Simple IIR low pass filter: H(z)=0.5z^-1/(1-0.5z^-1)
+        nps_ = 0.5 * (nps + nps_);
```

Comment:
This formula to me looks like a it just slows the rate of change of the nps estimate value.  Whether the new nps value is lower or higher than before, its replaced with the average.

Since the nps isn't preserved across moves (since a new search instance is created for every move) I don't see why this would result in an improvement.  My understanding is that testing shows that early nps estimates, even after accounting for the tolerances, are never actually too large - so this will only slow the convergence to the actual long term nps value by keeping the value too low.

Reply:
Some tests show it to be an improvement in a system with a periodic heavy load: blas backend and simultaneous integer programming solver using all threads for 1 min and off for another 1,

However, it also seems to be leaning towards being negative under normal conditions, so no joy.

Reply:
@Tilps: are you sure early nps estimates are always below the long term value? Maybe this is the case for the UCI reported value, which is different from the formula used for smart pruning:
`auto nps = (1000LL * total_playouts_ + kSmartPruningToleranceNodes) /                  (time_since_start - kSmartPruningToleranceMs) + 1;`

We divide the total number of nodes since the start of search (**including** `kSmartPruningToleranceMs` interval) + `kSmartPruningToleranceNodes` by the time **since** `kSmartPruningToleranceMs`.

Hence, to me, this seems that if the tolerance window is sufficiently large to allow transient responses to disappear, the early nps estimates will be higher than the actual long term nps value.

Reply:
the tolerance nodes is mistakenly outside of the multiply by 1000, so its actually 0.1 nodes of tolerance.
and the initial batches can be so slow that the first time you sample past the tolerance in time, its not sufficient to offset it.  Now there are probably times when the first sample past the tolerance in time is really close, and that gives it a pretty large nudge upwards - but I'm pretty sure you would find that most times the first samples past the initial tolerance are indeed below the long term NPS.